[XUnitLogChecker] Fix condition that sets env variables for libraries runner scripts

carlossanlop opened the pull request:
`$(BuildTargetFramework)`  was showing up empty at the point where it was needed.
I changed it to an existing condition I found in the same file to ensure we have the right target framework.
I also moved the `SetScriptCommands` itemgroup inside the target that consumes it, to keep it consistent with the other similar itemgroup for `RunScriptCommands`.

carlossanlop committed:
tmp

carlossanlop committed:
[XUnitLogChecker] Fix condition that sets env variables for libraries runner scripts.

ericstj commented on eng/testing/tests.targets:
```diff
@@ -136,6 +132,23 @@
       <RunScriptOutputPath>$([MSBuild]::NormalizePath('$(RunScriptOutputDirectory)', '$(RunScriptOutputName)'))</RunScriptOutputPath>
     </PropertyGroup>
 
+    <PropertyGroup Condition="'$(IsXUnitLogCheckerSupported)' == 'true' and
```
I see you switched to use a property for consistency with `RunScriptCommand`.  I think with `RunScriptCommand` that was used to represent a single command to run the test whereas with set it's meant to be a list.  IMO it'd be more readable to keep `SetScriptCommands` as an item:

```xml
<ItemGroup Condition="'$(IsXUnitLogCheckerSupported)' == 'true' and
                      '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
                      $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '$(NETCoreAppCurrentVersion)'))">
  <SetScriptCommands Condition="'$(TargetOS)' == 'windows'" Include="set __TestArchitecture=$(TargetArchitecture);set __IsXUnitLogCheckerSupported=1" />
  <SetScriptCommands Condition="'$(TargetOS)' != 'windows'" Include="export __TestArchitecture=$(TargetArchitecture);export __IsXUnitLogCheckerSupported=1" />
</ItemGroup>
```

PS: I noticed this when asking the questions: who else sets `SetScriptCommand`?  Will it be semi-colon, or new-line delimited?

ivdiazsa replied:
I think we should keep `SetScriptCommands` as an `ItemGroup`. It's also used by `tests.browser.targets`, `tests.wasi.targets`, and the `Directory.Build.targets` in `src/mono/sample`. It might not cause any issues whatsoever, but I'm scared it could potentially collide with them in the future, as they all treat it as a list in an `ItemGroup`.

carlossanlop replied:
>  who else sets SetScriptCommand?

In tests.targets? No one else was using SetScriptCommand. I added its use first use here: https://github.com/dotnet/runtime/pull/94868/files#diff-ea069e899aadcfcf3dcabecd88dbd80e12d6cb331de055fba7f49c9ef9a329adR72

I also added its first use in the runner templates used by libraries: [cmd](https://github.com/dotnet/runtime/pull/94868/files#diff-0bcef275f5c5df0bbd24b8db87c2d4c65cecb2770abe0463dac97074d19f6d5aR71-R72) and [sh](https://github.com/dotnet/runtime/pull/94868/files#diff-f7fb37056cce8903c2a743e9d9f9d6b9eee30041677827b9e0fa51d5f9378012R167-R168).

As Ivan mentioned, it is already used in many other places.

Just in case, that's why I made sure to precede the value of the properties with `$(SetScriptCommand)` in case there is a pre-existing value. I used a similar example from lines 99 and 100.

I also decided to follow the same pre-existing pattern of setting the additional values inside a property (with the name in singular) where I prefix the value with the same property, then include it to the item in a new itemgroup (with the name in plural), and then use the itemgroup in the task attribute (see RunScriptCommand and RunScriptCommands).

carlossanlop replied:
> Will it be semi-colon, or new-line delimited?

I have noticed both uses in several files. Seems to be the same, although I got a suggestion in my previous PR (in a chat conversation, not as PR feedback) to use endlines instead of semicolons. Both seem to work fine.

carlossanlop committed:
Revert the crash added on purpose to confirm the tool works. Commit 375f940425da210699ad27872a6a6e532a1fcf62.

carlossanlop commented:
I addressed the feedback with answers to the questions. I also reverted the crash. Can I get a sign-off if there is no blocking feedback? @ericstj @ivdiazsa

ViktorHofer commented on eng/testing/tests.targets:
```diff
@@ -136,6 +132,23 @@
+    <PropertyGroup Condition="'$(IsXUnitLogCheckerSupported)' == 'true' and
+                              '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
+                              $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '$(NETCoreAppCurrentVersion)'))">
+      <SetScriptCommand Condition="'$(TargetOS)' == 'windows'">$(SetScriptCommand)
+set __TestArchitecture=$(TargetArchitecture);
+set __IsXUnitLogCheckerSupported=1
+      </SetScriptCommand>
+      <SetScriptCommand Condition="'$(TargetOS)' != 'windows'">$(SetScriptCommand)
+export __TestArchitecture=$(TargetArchitecture);
+export __IsXUnitLogCheckerSupported=1
+      </SetScriptCommand>        
```
These should be items instead of properties.

carlossanlop replied:
Ok sure, I can use items. Note that there's lots of examples, especially with helix-related data, where we use a property first to collect all the elements, and then store them into a single item.

ViktorHofer replied:
This isn't necessary here. I.e. here you can just inline the value instead of defining a separate property:

```
<PropertyGroup>
  <CompileItem>A.cs</CompileItem>
</PropertyGroup>

<ItemGroup>
  <Compile Include="$(CompileItem)" />
</ItemGroup>
```

carlossanlop replied:
Pushed the commit.

carlossanlop replied:
I confirmed the generated runner script contains the env variables as expected.

carlossanlop committed:
Use items instead of property

ViktorHofer commented on eng/testing/tests.targets:
```diff
@@ -136,6 +132,16 @@
     </PropertyGroup>
 
+    <ItemGroup Condition="'$(IsXUnitLogCheckerSupported)' == 'true' and
+                              '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
+                              $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '$(NETCoreAppCurrentVersion)'))">
```
nit: indentation is wrong

ViktorHofer commented on eng/testing/tests.targets:
```diff
@@ -136,6 +132,16 @@
       <RunScriptOutputPath>$([MSBuild]::NormalizePath('$(RunScriptOutputDirectory)', '$(RunScriptOutputName)'))</RunScriptOutputPath>
     </PropertyGroup>
 
+    <ItemGroup Condition="'$(IsXUnitLogCheckerSupported)' == 'true' and
+                              '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
+                              $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '$(NETCoreAppCurrentVersion)'))">
+      <SetScriptCommands Condition="'$(TargetOS)' == 'windows'" Include="set __TestArchitecture=$(TargetArchitecture)" />
+      <SetScriptCommands Condition="'$(TargetOS)' == 'windows'" Include="set __IsXUnitLogCheckerSupported=1" />
+      
+      <SetScriptCommands Condition="'$(TargetOS)' != 'windows'" Include="export __TestArchitecture=$(TargetArchitecture)" />
+      <SetScriptCommands Condition="'$(TargetOS)' != 'windows'" Include="export __IsXUnitLogCheckerSupported=1" />
+    </ItemGroup>
```
What's the reason that you moved this block into the target? I would prefer it would stay outside of it. The target itself just calculates inputs and outputs and invokes the msbuild task while this block is proprietary to the xunit log checker.

carlossanlop replied:
Because I noticed RunScriptCommands is inside the target, and it is consumed exactly like SetScriptCommands. So I wanted to keep both consistent.

carlossanlop replied:
This is what I mean: https://github.com/dotnet/runtime/pull/95533/files#diff-ea069e899aadcfcf3dcabecd88dbd80e12d6cb331de055fba7f49c9ef9a329adR98-R128

Also, notice how it collects everything into a property, then puts it in an item.

ViktorHofer replied:
I think you refer to L124/L129. That is just putting the RunScriptCommand property into an item, so just build infrastructure for task inputs. The whole target is a general purpose target, not specific to a custom command. Please revert this change and put this back outside of the target.

(FWIW we chose a property back then because the majority of use cases only required a single command. We probably should have just used the item instead.)

carlossanlop committed:
Fix indentation

carlossanlop commented on eng/testing/tests.targets:
```diff
@@ -107,6 +102,7 @@
 
       <!-- Escape potential user input. -->
       <RunScriptCommand>$([MSBuild]::Escape('$(RunScriptCommand)'))</RunScriptCommand>
+
```
```suggestion
```

carlossanlop committed:
Remove extra space

carlossanlop committed:
Move items outside of task
